pkg/test/spaceprovisionerconfig/spaceprovisionerconfig_assertions.go

```diff
@@ -8,81 +8,132 @@ import (
-	notReady           struct{}
-	notReadyWithReason struct {
-		expectedReason string
+	readyWithStatusAndReason struct {
+		expectedStatus corev1.ConditionStatus
+		expectedReason *string
 	}
+
+	consumedSpaceCount struct {
+		expectedSpaceCount int
+	}
+
+	consumedMemoryUsage struct {
+		expectedMemoryUsage map[string]int
+	}
+
+	unknownConsumedCapacity struct{}
 )
 
 var (
-	_ assertions.Predicate[*toolchainv1alpha1.SpaceProvisionerConfig]           = (*ready)(nil)
-	_ assertions.Predicate[*toolchainv1alpha1.SpaceProvisionerConfig]           = (*notReady)(nil)
-	_ assertions.Predicate[*toolchainv1alpha1.SpaceProvisionerConfig]           = (*notReadyWithReason)(nil)
-	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*ready)(nil)
-	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*notReady)(nil)
-	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*notReadyWithReason)(nil)
+	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*readyWithStatusAndReason)(nil)
+	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*consumedSpaceCount)(nil)
+	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*consumedMemoryUsage)(nil)
+	_ assertions.PredicateMatchFixer[*toolchainv1alpha1.SpaceProvisionerConfig] = (*unknownConsumedCapacity)(nil)
 )
 
-func (*ready) Matches(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
-	return condition.IsTrueWithReason(spc.Status.Conditions, toolchainv1alpha1.ConditionReady, toolchainv1alpha1.SpaceProvisionerConfigValidReason)
-}
+func (r *readyWithStatusAndReason) Matches(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
+	cond, found := condition.FindConditionByType(spc.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	if !found {
+		return false
+	}
 
-func (*ready) FixToMatch(spc *toolchainv1alpha1.SpaceProvisionerConfig) *toolchainv1alpha1.SpaceProvisionerConfig {
-	spc.Status.Conditions, _ = condition.AddOrUpdateStatusConditions(spc.Status.Conditions, toolchainv1alpha1.Condition{
-		Type:   toolchainv1alpha1.ConditionReady,
-		Status: corev1.ConditionTrue,
-		Reason: toolchainv1alpha1.SpaceProvisionerConfigValidReason,
-	})
-	return spc
-}
+	if cond.Status != r.expectedStatus {
+		return false
+	}
 
-func Ready() assertions.Predicate[*toolchainv1alpha1.SpaceProvisionerConfig] {
-	return &ready{}
-}
+	if r.expectedReason != nil && cond.Reason != *r.expectedReason {
+		return false
+	}
 
-func (*notReady) Matches(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
-	return condition.IsFalse(spc.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	return true
 }
 
-func (*notReady) FixToMatch(spc *toolchainv1alpha1.SpaceProvisionerConfig) *toolchainv1alpha1.SpaceProvisionerConfig {
+func (r *readyWithStatusAndReason) FixToMatch(spc *toolchainv1alpha1.SpaceProvisionerConfig) *toolchainv1alpha1.SpaceProvisionerConfig {
```

Comment:
`FixToMatch` is a method of the `PredicateMatchFixer` interface https://github.com/codeready-toolchain/toolchain-common/blob/master/pkg/test/assertions/predicates.go#L50. Through the `Explain` function https://github.com/codeready-toolchain/toolchain-common/blob/master/pkg/test/assertions/assertions.go#L73, it is used in `AssertThat` in the same file and in the `wait.For()...` assertions in the e2e.

It's used to produce diffs that show what the tests expected the object to look like.

Granted, it's not used much because most of the existing tests weren't written with it in mind, but it's used in SPC tests and also in some of the NSTemplateTier tests that @mfrancisc is writing.

Reply:
ok, I wish we could have something more generic. 
It looks like that to be able to compare and print the diff, you need to always have the types defined for the expected values, which doesn't seem to be flexible. 
Anyway, I don't want to block this PR because of that, but let's discuss the options later.